synapse/storage/_base.py

```diff
@@ -192,6 +192,41 @@ def __init__(self, db_conn, hs):
 
         self.database_engine = hs.database_engine
 
+        # A set of tables that are not safe to use native upserts in.
+        self._unsafe_to_upsert_tables = {"user_ips"}
```

Comment:
Presumably the idea here is that we add to this whenever we add an index that we may want to upsert against? This feels fairly fragile tbh, since it'll probably work in CI and then break in production. I wonder if we can hook into the `register_background_index_update` to add to this list when adding a unique index?

Its possibly fine for now

Reply:
I ran this past @richvdh, and broadly the conclusion is that this should be an edge case (since we'll aim to add unique indices from the start to new tables), so we can leave it for now. (https://matrix.to/#/!yZHTGeDKZUeKaqeTeU:matrix.org/$1548086288327WrmxW:jki.re?via=matrix.org&via=half-shot.uk&via=sw1v.org is the conversation)